transport_plugins/bled112/iotile_transport_bled112/async_packet.py

```diff
@@ -51,37 +60,39 @@ def read_packet(self, timeout=3.0):
             raise InternalTimeoutError("Timeout waiting for packet in AsyncPacketBuffer")
 
 
-def ReaderThread(filelike, read_queue, header_length, length_function, stop):
+def reader_thread(filelike, read_queue, header_length, length_function, stop, dedupe=False, dedupe_timeout=5):
     logger = logging.getLogger(__name__)
+    broadcast_v2_dedupers = None
+    read_buffer = bytearray()
+
+    if dedupe:
+        broadcast_v2_dedupers = BroadcastV2DeduperCollection(dedupe_timeout)
 
     while not stop.is_set():
         try:
-            header = bytearray()
-            while len(header) < header_length:
-                chunk = bytearray(filelike.read(header_length - len(header)))
-                header += chunk
+            # The bled112 will read EOF when there is no more data, so it's safe to read a large amount
+            read_buffer += bytearray(filelike.read(1024))
 
-                if stop.is_set():
+            while not stop.is_set() and len(read_buffer) >= header_length:
+                next_packet_len = header_length + length_function(read_buffer[:header_length])
+
+                if len(read_buffer) < next_packet_len:
+                    # Still waiting to read this packet
                     break
 
-            if stop.is_set():
-                break
+                # Process the packet and remove it from the read buffer
+                packet = read_buffer[:next_packet_len]
+                del read_buffer[:next_packet_len]
```

Comment:
I don't know for sure that this is bad but I've not seen this pattern before for doing a `tail` on a buffer.  Normally I see:

```
read_buffer = read_buffer[next_packet_len:]
```

Reply:
I didn't see any apparent issues from it, but I can change it.